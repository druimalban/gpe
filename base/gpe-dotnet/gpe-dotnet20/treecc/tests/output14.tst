// test skeleton "yy" replacement

%option prefix = "xyzzy"

%node expression %abstract %typedef
