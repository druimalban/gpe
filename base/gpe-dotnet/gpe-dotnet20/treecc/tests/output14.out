line 3: %option prefix xyzzy
line 5: %node expression no_parent 6
/* output.h.  Generated automatically by treecc */
#ifndef __xyzzy_output_h
#define __xyzzy_output_h

#ifdef __cplusplus
extern "C" {
#endif

#define expression_kind 1

typedef struct expression__ expression;

#line 1 "c_skel.h"
typedef struct
{
	struct YYNODESTATE_block *blocks__;
	struct YYNODESTATE_push *push_stack__;
	int used__;

} YYNODESTATE;
#line 22 "output.h"
struct expression__ {
	const struct expression_vtable__ *vtable__;
	int kind__;
	char *filename__;
	long linenum__;
};

struct expression_vtable__ {
	const void *parent__;
	int kind__;
	const char *name__;
};

extern struct expression_vtable__ const expression_vt__;



#ifndef xyzzykind
#define xyzzykind(node__) ((node__)->kind__)
#endif

#ifndef xyzzykindname
#define xyzzykindname(node__) ((node__)->vtable__->name__)
#endif

#ifndef xyzzykindof
#define xyzzykindof(type__) (type__##_kind)
#endif

#ifndef xyzzyisa
extern int xyzzyisa__(const void *vtable__, int kind__);
#define xyzzyisa(node__,type__) \
	(xyzzyisa__((node__)->vtable__, (type__##_kind)))
#endif

#ifndef xyzzygetfilename
#define xyzzygetfilename(node__) ((node__)->filename__)
#endif

#ifndef xyzzygetlinenum
#define xyzzygetlinenum(node__) ((node__)->linenum__)
#endif

#ifndef xyzzysetfilename
#define xyzzysetfilename(node__, value__) \
	((node__)->filename__ = (value__))
#endif

#ifndef xyzzysetlinenum
#define xyzzysetlinenum(node__, value__) \
	((node__)->linenum__ = (value__))
#endif

#ifndef xyzzytracklines_declared
extern char *xyzzycurrfilename(void);
extern long xyzzycurrlinenum(void);
#define xyzzytracklines_declared 1
#endif

#ifndef xyzzynodeops_declared
extern void xyzzynodeinit(void);
extern void *xyzzynodealloc(unsigned int size__);
extern int xyzzynodepush(void);
extern void xyzzynodepop(void);
extern void xyzzynodeclear(void);
extern void xyzzynodefailed(void);
#define xyzzynodeops_declared 1
#endif

#ifdef __cplusplus
};
#endif

#endif
/* output.c.  Generated automatically by treecc */

#define YYNODESTATE_TRACK_LINES 1
#line 1 "c_skel.c"
/*
 * treecc node allocation routines for C.
 *
 * Copyright (C) 2001  Southern Storm Software, Pty Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * As a special exception, when this file is copied by treecc into
 * a treecc output file, you may use that output file without restriction.
 */

#include <stdlib.h>

#ifndef YYNODESTATE_BLKSIZ
#define	YYNODESTATE_BLKSIZ	2048
#endif

/*
 * Types used by the allocation routines.
 */
struct YYNODESTATE_block
{
	char data__[YYNODESTATE_BLKSIZ];
	struct YYNODESTATE_block *next__;

};
struct YYNODESTATE_push
{
	struct YYNODESTATE_push *next__;
	struct YYNODESTATE_block *saved_block__;
	int saved_used__;
};

/*
 * The fixed global state to use for non-reentrant allocation.
 */
#ifndef YYNODESTATE_REENTRANT
static YYNODESTATE fixed_state__;
#endif

/*
 * Some macro magic to determine the default alignment
 * on this machine.  This will compile down to a constant.
 */
#define	YYNODESTATE_ALIGN_CHECK_TYPE(type,name)	\
	struct _YYNODESTATE_align_##name { \
		char pad; \
		type field; \
	}
#define	YYNODESTATE_ALIGN_FOR_TYPE(type)	\
	((unsigned)(&(((struct _YYNODESTATE_align_##type *)0)->field)))
#define	YYNODESTATE_ALIGN_MAX(a,b)	\
	((a) > (b) ? (a) : (b))
#define	YYNODESTATE_ALIGN_MAX3(a,b,c) \
	(YYNODESTATE_ALIGN_MAX((a), YYNODESTATE_ALIGN_MAX((b), (c))))
YYNODESTATE_ALIGN_CHECK_TYPE(int, int);
YYNODESTATE_ALIGN_CHECK_TYPE(long, long);
#if defined(WIN32) && !defined(__CYGWIN__)
YYNODESTATE_ALIGN_CHECK_TYPE(__int64, long_long);
#else
YYNODESTATE_ALIGN_CHECK_TYPE(long long, long_long);
#endif
YYNODESTATE_ALIGN_CHECK_TYPE(void *, void_p);
YYNODESTATE_ALIGN_CHECK_TYPE(float, float);
YYNODESTATE_ALIGN_CHECK_TYPE(double, double);
#define	YYNODESTATE_ALIGNMENT	\
	YYNODESTATE_ALIGN_MAX( \
			YYNODESTATE_ALIGN_MAX3	\
			(YYNODESTATE_ALIGN_FOR_TYPE(int), \
		     YYNODESTATE_ALIGN_FOR_TYPE(long), \
			 YYNODESTATE_ALIGN_FOR_TYPE(long_long)), \
  	     YYNODESTATE_ALIGN_MAX3 \
		 	(YYNODESTATE_ALIGN_FOR_TYPE(void_p), \
			 YYNODESTATE_ALIGN_FOR_TYPE(float), \
			 YYNODESTATE_ALIGN_FOR_TYPE(double)))

/*
 * Initialize the node allocation pool.
 */
#ifdef YYNODESTATE_REENTRANT
void xyzzynodeinit(state__)
YYNODESTATE *state__;
{
#else
void xyzzynodeinit()
{
	YYNODESTATE *state__ = &fixed_state__;
#endif
	state__->blocks__ = 0;
	state__->push_stack__ = 0;
	state__->used__ = 0;
}

/*
 * Allocate a block of memory.
 */
#ifdef YYNODESTATE_REENTRANT
void *xyzzynodealloc(state__, size__)
YYNODESTATE *state__;
unsigned int size__;
{
#else
void *xyzzynodealloc(size__)
unsigned int size__;
{
	YYNODESTATE *state__ = &fixed_state__;
#endif
	struct YYNODESTATE_block *block__;
	void *result__;

	/* Round the size to the next alignment boundary */
	size__ = (size__ + YYNODESTATE_ALIGNMENT - 1) &
				~(YYNODESTATE_ALIGNMENT - 1);

	/* Do we need to allocate a new block? */
	block__ = state__->blocks__;
	if(!block__ || (state__->used__ + size__) > YYNODESTATE_BLKSIZ)
	{
		if(size__ > YYNODESTATE_BLKSIZ)
		{
			/* The allocation is too big for the node pool */
			return (void *)0;
		}
		block__ = (struct YYNODESTATE_block *)
						malloc(sizeof(struct YYNODESTATE_block));
		if(!block__)
		{
			/* The system is out of memory.  The programmer can
			   supply the "xyzzynodefailed" function to report the
			   out of memory state and/or abort the program */
#ifdef YYNODESTATE_REENTRANT
			xyzzynodefailed(state__);
#else
			xyzzynodefailed();
#endif
			return (void *)0;
		}
		block__->next__ = state__->blocks__;
		state__->blocks__ = block__;
		state__->used__ = 0;
	}

	/* Allocate the memory and return it */
	result__ = (void *)(block__->data__ + state__->used__);
	state__->used__ += size__;
	return result__;
}

/*
 * Push the node allocation state.
 */
#ifdef YYNODESTATE_REENTRANT
int xyzzynodepush(state__)
YYNODESTATE *state__;
{
#else
int xyzzynodepush()
{
	YYNODESTATE *state__ = &fixed_state__;
#endif
	struct YYNODESTATE_block *saved_block__;
	int saved_used__;
	struct YYNODESTATE_push *push_item__;

	/* Save the current state of the node allocation pool */
	saved_block__ = state__->blocks__;
	saved_used__ = state__->used__;

	/* Allocate space for a push item */
#ifdef YYNODESTATE_REENTRANT
	push_item__ = (struct YYNODESTATE_push *)
			xyzzynodealloc(state__, sizeof(struct YYNODESTATE_push));
#else
	push_item__ = (struct YYNODESTATE_push *)
			xyzzynodealloc(sizeof(struct YYNODESTATE_push));
#endif
	if(!push_item__)
	{
		return 0;
	}

	/* Copy the saved information to the push item */
	push_item__->saved_block__ = saved_block__;
	push_item__->saved_used__ = saved_used__;

	/* Add the push item to the push stack */
	push_item__->next__ = state__->push_stack__;
	state__->push_stack__ = push_item__;
	return 1;
}

/*
 * Pop the node allocation state.
 */
#ifdef YYNODESTATE_REENTRANT
void xyzzynodepop(state__)
YYNODESTATE *state__;
{
#else
void xyzzynodepop()
{
	YYNODESTATE *state__ = &fixed_state__;
#endif
	struct YYNODESTATE_push *push_item__;
	struct YYNODESTATE_block *saved_block__;
	struct YYNODESTATE_block *temp_block__;

	/* Pop the top of the push stack */
	push_item__ = state__->push_stack__;
	if(push_item__ == 0)
	{
		saved_block__ = 0;
		state__->used__ = 0;
	}
	else
	{
		saved_block__ = push_item__->saved_block__;
		state__->used__ = push_item__->saved_used__;
		state__->push_stack__ = push_item__->next__;
	}

	/* Free unnecessary blocks */
	while(state__->blocks__ != saved_block__)
	{
		temp_block__ = state__->blocks__;
		state__->blocks__ = temp_block__->next__;
		free(temp_block__);
	}
}

/*
 * Clear the node allocation pool completely.
 */
#ifdef YYNODESTATE_REENTRANT
void xyzzynodeclear(state__)
YYNODESTATE *state__;
{
#else
void xyzzynodeclear()
{
	YYNODESTATE *state__ = &fixed_state__;
#endif
	struct YYNODESTATE_block *temp_block__;
	while(state__->blocks__ != 0)
	{
		temp_block__ = state__->blocks__;
		state__->blocks__ = temp_block__->next__;
		free(temp_block__);
	}
	state__->push_stack__ = 0;
	state__->used__ = 0;
}
#line 270 "output.c"
struct expression_vtable__ const expression_vt__ = {
	0,
	expression_kind,
	"expression",
};

struct xyzzy_vtable__ {
	const struct xyzzy_vtable__ *parent__;
	int kind__;
};

int xyzzyisa__(const void *vtable__, int kind__)
{
	const struct xyzzy_vtable__ *vt;
	vt = (const struct xyzzy_vtable__ *)vtable__;
	while(vt != 0) {
		if(vt->kind__ == kind__)
			return 1;
		vt = vt->parent__;
	}
	return 0;
}

